Logic/igtlioConnector.cxx

```diff
@@ -707,85 +730,229 @@ void igtlioConnector::PushOutgoingMessages()
-      }
+      this->PeriodicProcess();
+      vtksys::SystemTools::Delay(5);
     }
+  }
+  else
+  {
+    vtkErrorMacro("Could not send command " << command->GetCommandId() << ": \"" << command->GetName() << "\"to client");
+    command->SetStatus(igtlioCommandStatus::CommandFailed);
+  }
+  return success;
+}
 
-  igtlioCommandConverter::ContentData contentdata = device->GetContent();
-  contentdata.id += this->NextCommandID++;
-  contentdata.name = command;
-  contentdata.content = content;
-  device->SetContent(contentdata);
-  device->SetQueryTimeOut(timeout_s);
+//----------------------------------------------------------------------------
+int igtlioConnector::SendCommandResponse(int commandId, int clientId/*=-1*/)
+{
+  igtlioCommandPointer response = igtlioCommandPointer::New();
+  response->SetCommandId(commandId);
 
-  device->PruneCompletedQueries();
+  return this->SendCommandResponse(response);
+}
 
-  SendMessage(igtlioDeviceKeyType::CreateDeviceKey(device));
+//----------------------------------------------------------------------------
+int igtlioConnector::SendCommandResponse(igtlioCommandPointer command)
+{
 
-  return device;
+  if (command->GetStatus() != igtlioCommandStatus::CommandWaiting)
+    {
+    return 0;
+    }
+
+  igtl::RTSCommandMessage::Pointer responseMessage = igtl::RTSCommandMessage::New();
+  igtlioBaseConverter::HeaderData headerData = igtlioBaseConverter::HeaderData();
+  igtlioCommandConverter::ContentData content = igtlioCommandConverter::ContentData();
+  content.id = command->GetCommandId();
+  content.name = command->GetName();
+  content.content = command->GetResponseContent();
+
+  igtl::MessageBase::MetaDataMap metaData = command->GetResponseMetaData();
+  igtl::CommandMessage::Pointer commandMessage = responseMessage;
+  igtlioCommandConverter::toIGTL(headerData, content, &commandMessage, metaData);
+
+  int success = this->SendData(responseMessage->GetPackSize(), (unsigned char*)responseMessage->GetPackPointer());
+  command->SetStatus(igtlioCommandStatus::CommandResponseSent);
+  return success;
+}
+
+//----------------------------------------------------------------------------
+void igtlioConnector::CancelCommand(int commandId, int clientId/*=-1*/)
+{
+  igtlioCommandPointer command = this->GetOutgoingCommand(commandId);
+  this->CancelCommand(command);
+}
+
+//----------------------------------------------------------------------------
+void igtlioConnector::CancelCommand(igtlioCommandPointer command)
+{
+  if (!command)
+  {
+    return;
+  }
+
+  command->SetStatus(igtlioCommandStatus::CommandCancelled);
+  this->PruneCompletedCommands();
+
+  this->InvokeEvent(igtlioCommand::CommandCancelledEvent, command.GetPointer());
+  command->InvokeEvent(igtlioCommand::CommandCancelledEvent);
 }
 
 //----------------------------------------------------------------------------
 void igtlioConnector::ParseCommands()
 {
-  this->CommandQueueMutex->Lock();
+  this->IncomingCommandQueueMutex->Lock();
```

Comment:
Does this mutex class have an associated lock object? Something like `xyzLockGuard(this->IncomingCommandQueueMutex)`?

Reply:
IncomingCommandQueueMutex is just a vtkSmartPointer<vtkMutexLock>. Are there any such LockGuards in VTK?

Reply:
We could bring over PlusLockGuard, it is about 15 lines of code.

Reply:
Sure, I can add it to igtlioUtilities as igtlioLockGuard unless there is a better place/name

Reply:
Done